### PR TITLE
Row outline/hidden/collapsed attributes support (xlsx only)

### PR DESCRIPTION
**Pull request by LastDragon-ru**
image

### CODE REVIEW

File: docs/documentation.md [hunk 1]
```diff
@@ -181,6 +181,39 @@ $options->mergeCells(0, 1, 0, 2, $writer->getCurrentSheet()->getIndex());
...
+use OpenSpout\Writer\XLSX\Writer;
+use OpenSpout\Writer\XLSX\Options;
+use OpenSpout\Writer\XLSX\RowAttributes;
+
+$options = new Options();
+
+// Required or the generated file will be corrupted
+// Looks like empty rows/cell cannot have `defaultRowHeight` equal to 0
+// which will be added when the setting is not set.
+$options->DEFAULT_ROW_HEIGHT = 15;
```
Comment (Slamdunk) [inline docs/documentation.md:214]: Can we something about this, and not require the user to specify it?

File: src/Writer/XLSX/Manager/WorksheetManager.php [hunk 2]
```diff
@@ -142,6 +149,56 @@ private function addNonEmptyRow(Worksheet $worksheet, Row $row): void
...
+        $rowAttributes = ''
+            ." r=\"{$rowIndexOneBased}\""
+            .$this->getRowAttributes($row);
+
+        $rowXML = "<row{$rowAttributes}></row>";
+
+        $wasWriteSuccessful = fwrite($sheetFilePointer, $rowXML);
+        if (false === $wasWriteSuccessful) {
+            throw new IOException("Unable to write data in {$worksheet->getFilePath()}");
+        }
```
Comment (Slamdunk) [inline src/Writer/XLSX/Manager/WorksheetManager.php]: I see you've just copy-pasted this and can't be tested: I say let's just assert it:
```suggestion
        assert(false !== $wasWriteSuccessful);
```

File: src/Writer/XLSX/Options.php [hunk 3]
```diff
@@ -63,4 +70,22 @@ public function getMergeCells(): array
...
+        $this->rowsAttributes[$row] = $attributes;
+        $this->outlineMaxLevel = max($attributes->getOutlineLevel(), (int) $this->outlineMaxLevel);
+
+        return $row;
+    }
+
+    public function getRowAttributes(Row $row): ?RowAttributes
+    {
+        return $this->rowsAttributes[$row] ?? null;
+    }
```
Comment (Slamdunk) [inline src/Writer/XLSX/Options.php]: Nullable elements only leads to weak types; let's just return an empty `RowAttributes` and implement an `isEmpty` method on that object
```suggestion
    public function getRowAttributes(Row $row): RowAttributes
    {
        return $this->rowsAttributes[$row] ?? new RowAttributes();
    }
```

File: src/Writer/XLSX/RowAttributes.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenSpout\Writer\XLSX;
+
+use OpenSpout\Common\Exception\InvalidArgumentException;
+
+class RowAttributes
```
Comment (Slamdunk) [inline src/Writer/XLSX/RowAttributes.php]:
```suggestion
final class RowAttributes
```

File: src/Writer/XLSX/RowAttributes.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    private bool $visible;
+    private bool $collapsed;
+
+    /** @var null|int<0, 7> */
+    private ?int $outlineLevel;
+
+    /**
+     * @param null|int<0, 7> $outlineLevel
+     */
+    public function __construct(?int $outlineLevel, bool $collapsed = false, bool $visible = true)
```
Comment (Slamdunk) [inline src/Writer/XLSX/RowAttributes.php]: This whole class can just be a Value Object. Let's remove PHP `8.0.0` support and add an Enum for the OutlineLevel, then we can rewrite this class as:
```php
enum OutlineLevel: int
{
    case ONE    = 1;
    case TWO    = 2;
    case THREE  = 3;
    case FOUR   = 4;
    case FIVE   = 5;
    case SIX    = 6;
    case SEVEN  = 7;
}

final class RowAttributes
{
    public function __construct(
        public readonly ?OutlineLevel $OUTLINE_LEVEL = null,
        public readonly bool $COLLAPSED = false,
        public readonly bool $VISIBLE = true,
    ) {}
    
    public function isEmpty(): bool
    {
        return null === $this->OUTLINE_LEVEL
            && false === $this->COLLAPSED
            && true === $this->VISIBLE
        ;
    }
}
```

File: src/Writer/XLSX/Options.php [hunk 6]
```diff
@@ -63,4 +70,22 @@ public function getMergeCells(): array
     {
         return $this->MERGE_CELLS;
     }
+
+    public function setRowAttributes(Row $row, RowAttributes $attributes): Row
```
Comment (Slamdunk) [inline src/Writer/XLSX/Options.php]: No fluent interfaces please
```suggestion
    public function setRowAttributes(Row $row, RowAttributes $attributes): void
```

**Review by Slamdunk — CHANGES_REQUESTED** [review]
This looks awesome! Just few things to tweak and then :ship:

File: src/Writer/XLSX/RowAttributes.php (hunk 5 above)
Comment (LastDragon-ru) [reply src/Writer/XLSX/RowAttributes.php]: > Let's remove PHP 8.0.0 support

No please, our project still on PHP 8.0)

File: docs/documentation.md (hunk 1 above)
Comment (LastDragon-ru) [reply docs/documentation.md:214]: The problem is in empty rows which are required to split outline regions (or Excell will merge them into one region) - with no height we will have `<sheetFormatPr defaultRowHeight="0"/>` and (web-based) Excel[^1][^2]  will report about cryptic error "not in a range" (or something like...) and file will not be opened at all. I've tried to add the cell, the cell with empty inline string, but no luck :(

Looks like height required in any case 🤷‍♂️

[^1]: standalone version seems can open it, but with "file was recovered" message
[^2]: Libreoffice Calc open file without any warnings

File: src/Writer/XLSX/RowAttributes.php (hunk 5 above)
Comment (Slamdunk) [reply src/Writer/XLSX/RowAttributes.php]: https://github.com/openspout/openspout/blob/4.x/UPGRADE.md#upgrading-from-3x-to-40

I'm gonna do it anyway in 14 days https://www.php.net/supported-versions.php and I won't be able to switch `$OUTLINE_LEVEL` type in a minor release, so let's just create it as an Enum.

Until you are on 8.0 you can download your PR, modify it and apply it with https://github.com/cweagans/composer-patches to the latest `openspout:4.8` release

File: docs/documentation.md (hunk 1 above)
Comment (Slamdunk) [reply docs/documentation.md:214]: I don't like the idea that this library can produce a corrupted file if a particular option is not set.

Can you have a try to have a workaround within the code? I.E. if an emtpy row is added, the default row height gets set to a non zero value?

File: src/Writer/XLSX/RowAttributes.php (hunk 5 above)
Comment (LastDragon-ru) [reply src/Writer/XLSX/RowAttributes.php]: Oh, I see... Unfortunately, I have only PHP 8.0 and cannot do it, sorry.

File: docs/documentation.md (hunk 1 above)
Comment (LastDragon-ru) [reply docs/documentation.md:214]: > `<row r="5" outlineLevel="1" customHeight="1" ht="15"></row>`

Doesn't work 😞 So I have no more ideas...

image

**Comment by Slamdunk** [comment]
@LastDragon-ru I'm converting this to `Draft`: feel free to ping me when you figured out how to get it working

**Comment by JonBaron78** [comment]
@LastDragon-ru How about a fallback to 15 like it is done in StyleManager::getContentXmlAutomaticStylesSectionContent() ?

`$defaultRowHeight = null === $this->options->DEFAULT_ROW_HEIGHT ? '15pt' : "{$this->options->DEFAULT_ROW_HEIGHT}pt";`

Maybe make fallback a constant?

**Comment by jhogervorst** [comment]
@LastDragon-ru @Slamdunk I found this PR, as I'm looking into using Openspout in a project, but we need the "collapsed" functionality. It looks like this PR is quite old.

What is the status actually? Are the two currently open conversations the only remaining things before this PR can be merged? Or are there more issues?

Maybe we can help completing this PR 🙂

**Comment by LastDragon-ru** [comment]
> What is the status actually?

@jhogervorst Unfortunately, I do not work on a project where this was used anymore, and have no time to update the PR. Sorry.

> Are the two currently open conversations the only remaining things before this PR can be merged?

As I understand/remember

1) Enum should be used instead `int` for level

2) Need some workaround to add empty row without `$options->DEFAULT_ROW_HEIGHT` setting by user, that will not crash ms excel and/or merge multiple regions.

**Comment by Slamdunk** [comment]
> Maybe we can help completing this PR

Yes please: open a new one though.

**Comment by jhogervorst** [comment]
> > Maybe we can help completing this PR
> 
> Yes please: open a new one though.

It turns out I made a mistake: for our project, we need to support collapsing _columns_, not _rows_.

So we'll not be completing this PR, but my colleague just opened a PR for collapsing columns: https://github.com/openspout/openspout/pull/275.

**Pull request closed without merging**